packages/core/src/authFlows/oidc_auth_code_flow.ts

```diff
@@ -0,0 +1,117 @@
+// Copyright 2020 Cognite AS
+import { LOCAL_STORAGE_PREFIX } from '../constants';
+import { UserManager, WebStorageStateStore } from 'oidc-client';
+
+export interface OIDCAuthFlowOptions {
+  clientId: string;
+  openIdConfigurationUrl: string;
+  responseMode: string;
+  responseType?: string;
+  extraScope?: string;
+  authenticateParams?: Record<string, any>;
+  refreshParams?: Record<string, any>;
+  loginParams?: Record<string, any>;
+
+  interactionType?: 'REDIRECT' | 'POPUP';
+}
+
+const localStorageKey = `${LOCAL_STORAGE_PREFIX}oidAuthCodeRedirect`;
+
+export class OidcAuthCode {
+  private readonly userManager: UserManager;
+  private readonly options: OIDCAuthFlowOptions;
+
+  constructor(options: OIDCAuthFlowOptions) {
+    this.options = options;
+    const {
+      openIdConfigurationUrl,
+      clientId,
+      responseMode,
+      responseType = 'code',
+      extraScope = '',
+      authenticateParams,
+    } = options;
+    this.userManager = new UserManager({
+      authority: openIdConfigurationUrl,
+      client_id: clientId,
+      response_mode: responseMode,
+      response_type: responseType,
+
+      scope: extraScope,
+      extraQueryParams: authenticateParams,
+      userStore: new WebStorageStateStore({ store: window.localStorage }),
+      loadUserInfo: false,
+
+      redirect_uri: `${window.location.protocol}//${window.location.host}`,
+    });
+  }
+
+  public async init() {
+    return this.userManager
+      .signinRedirectCallback()
+      .catch(e => {
+        if (
+          e.message === 'No state in response' ||
+          e.message === 'No matching state found in storage'
+        ) {
+          return this.userManager.getUser().then(u => {
+            if (u) {
+              if (u.expired) {
+                return this.userManager
+                  .signinSilent(this.options.refreshParams)
+                  .catch(() => Promise.resolve(null));
+              } else {
+                return Promise.resolve(u);
+              }
+            } else {
+              return Promise.resolve(null);
+            }
+          });
+        } else {
+          throw e;
+        }
+      })
+      .finally(() => {
+        const redirectStateStr = window.localStorage.getItem(localStorageKey);
```

Comment:
Just a suggestion here. Maybe it is good to write a wrapper class like storage service that will behave like a wrapper and just choose whatever storage type you need inside. Also this can do all the json parsing stuff.

Example:
```JS
export class StorageProvider {
private driver;

constructor(){ this.driver = window.localStorage; }

getItem() {
// check if it exists, do the json parsing and validation here

}
}

```

Reply:
please revisit the code. I removed the code from @vegardok s PR